V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-RESILIENCE/MASWE-0051.md

Reviewer: DONE

---

> title: Code Obfuscation Not Implemented

Reviewer: ```suggestion
title: Code Obfuscation Not Implemented In Security-Relevant Code
```

Reviewer 2: This one is debatable, since it's a 'security vs performance' issue. If you take this to extremes:
- A binary that's fully obfuscated doesn't give any obvious starting points and you could spend a lot of resources reversing a non-security critical function
- A binary with just one obfuscated function (e.g. the TLS pinning) gives an obvious starting point and focus point

If we keep the weakness generic, it would basically only fail if there's no code obfuscation at all. It's up to the tester to estimate how well the obfuscation has been performed.

Reviewer 2: + With this rename the focus would only be on security-relevant code, and not IP / business logic / etc

Reviewer: Makes sense to me, in other RESILIENCE weaknesses we have also focused on security-relevant code. Also in the "Overview" and "Modes of introduction" we already talk about "Security-relevant code". I can change the wording of the title/requirement and give a pass to this weakness so it focuses on security-relevant code. @TheDauntless

Reviewer 3: We keep it generic for now and also to be consistent with the wording for the other MASWE/Requirements and not include " In Security-Relevant Code". We can revisit this topic later after this is being merged as follow-up

id: MASWE-0051
alias: code-obfuscation

> requirement: "The app applies code obfuscation to hinder reverse engineering."

Reviewer: ```suggestion
requirement: "The app implements code obfuscation."
```

Reviewer 2: Same comment as ^

Reviewer 3: We keep it generic for now and also to be consistent with the wording for the other MASWE/Requirements and not include " In Security-Relevant Code". We can revisit this topic later after this is being merged as follow-up

platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0051
attacks: [MAS-ATTACK-0001]
mappings:
  masvs-v1: [MSTG-RESILIENCE-9]
  masvs-v2: [MASVS-RESILIENCE-3]
  cwe: [693]
  maswe-beta: [MASWE-0089, MASWE-0092]
refs:

> - https://developer.android.com/build/shrink-code

Reviewer: ```suggestion
- https://developer.android.com/topic/performance/app-optimization/enable-app-optimization#overview
```

status: new
---

## Overview

> This weakness occurs when an app's code, particularly its security-relevant logic, ships without effective obfuscation, making reverse engineering and static analysis straightforward.

Reviewer: ```suggestion
This weakness occurs when an app's code, particularly its security-relevant logic, ships without effective obfuscation, facilitating reverse engineering and static analysis.
```

> Obfuscation does not prevent reverse engineering, but it raises its cost. Effective schemes go beyond default identifier renaming and include techniques such as control-flow transformations, opaque predicates, instruction substitution, instruction block chopping, and method inlining, as well as hindering decompilers from producing usable output. Without them, attackers can quickly understand and locate the app's defenses and proprietary logic.

Reviewer: ```suggestion
Obfuscation does not prevent reverse engineering, but it raises its cost. Identifier renaming or string encryption are easier to implement and are considered common obfuscation techniques. However, there are other obfuscation techniques, like control-flow transformations, opaque predicates, instruction substitution, instruction block chopping, and method inlining, as well as hindering decompilers from producing usable output. Most of these techniques can be combined together.
```

## Modes of Introduction

- **No Obfuscation Applied**: Shipping release builds without any minification or obfuscation, leaving class, method, and symbol names intact.

> - **Security-Relevant Logic Left Readable**: Applying only default renaming while leaving security checks, licensing logic, and proprietary algorithms trivially analyzable, without stronger techniques such as opaque predicates, instruction substitution, or control-flow transformations.

Reviewer: ```suggestion
- **Security-Relevant Logic Left Readable**: Not applying enough obfuscation to security-relevant code, leaving code sections like security checks, licensing logic, and proprietary algorithms trivially analyzable.
```

Reviewer: I left two modes of introduction: not applying obfuscation at all, and applying obfuscation but not enough to make the security-relevant code readable. Maybe this can be only 1.

> - **Effectiveness Never Assessed**: Never testing the obfuscation against current decompilers and deobfuscation tooling, so its actual strength is unknown.

Reviewer: ```suggestion
```

## Impact

- **Bypass of Protection Mechanisms**: Attackers can quickly locate and defeat client-side checks such as root detection, licensing, or anti-tampering logic, resulting in the circumvention of the app's defenses.
- **Compromise of Sensitive Data**: Attackers can recover proprietary algorithms and embedded secrets from readable code, resulting in intellectual property theft and easier planning of further attacks.

## Mitigations

> - **Enable Obfuscation for Release Builds**: Apply the platform's minification and obfuscation tooling (e.g. R8) or established commercial obfuscators to all release builds.

Reviewer: ```suggestion
- **Enable Obfuscation for Release Builds**: Apply the platform's minification and obfuscation tooling or established commercial obfuscators to all release builds.
```

> - **Harden Security-Relevant Code**: Apply stronger transformations (opaque predicates, instruction substitution, control-flow obfuscation, method inlining) to the code implementing security checks and proprietary logic.

Reviewer: ```suggestion
- **Harden Security-Relevant Code**: Apply stronger transformations (opaque predicates, instruction substitution, control-flow obfuscation, method inlining) to the security-relevant code.
```

- **Assess Obfuscation Effectiveness**: Regularly attempt to reverse engineer the release build with state-of-the-art decompilers and deobfuscators to validate that the protection meets its goal.

> - **Combine with Runtime Protections**: Pair obfuscation with runtime integrity and anti-instrumentation checks (see @MASWE-0053, @MASWE-0061, @MASWE-0064), since obfuscation alone only delays attackers.

Reviewer: Is linking one MASWE with other proper? It seems fine but just to double check.

> - **Combine with Runtime Protections**: Pair obfuscation with runtime integrity and anti-instrumentation checks (see @MASWE-0053, @MASWE-0061, @MASWE-0064), since obfuscation alone only delays attackers.

Reviewer: ```suggestion
- **Combine with Runtime Protections**: Pair obfuscation with runtime integrity and anti-instrumentation checks (see the mitigations in @MASWE-0053, @MASWE-0061, @MASWE-0064), since obfuscation alone only delays attackers.
```
